src/Advection/weno_fifth_order.jl

```diff
@@ -1,134 +1,367 @@
+    β₀, β₁ and β₂ so that they account for the stretched `grid`. (defaults to `false`)
+  - `zweno`: When `true` implement a Z-WENO formulation for the WENO weights calculation. (defaults to
+    `false`)
+
+!!! warn "No support for WENO5 on curvilinear grids"
+    Currently, WENO 5th-order advection schemes don't work for for curvilinear grids.
+    Providing `WENO5(::AbstractCurvilinearGrid)` defaults to uniform setting, i.e.
+    `WENO5(::AbstractCurvilinearGrid) = WENO5()`.
+
+```@meta
+DocTestFilters = = [Regex("Warning: defaulting"), Regex("Oceananigans.Advection")"]
+```
+
+Not providing any keyword argument, `WENO5()` defaults to the uniform 5th-order coefficients ("uniform
+setting) in all directions, using a JS-WENO formulation.
+
+```jldoctest
+julia> using Oceananigans
+
+julia> WENO5()
+┌ Warning: defaulting to uniform WENO scheme with Float64 precision, use WENO5(grid = grid) if this was not intended
+└ @ Oceananigans.Advection ~/Research/OC2/src/Advection/weno_fifth_order.jl:90
+WENO5 advection scheme with:
+    ├── X regular
+    ├── Y regular
+    └── Z regular
+```
+
+`WENO5(grid = grid)` defaults to uniform interpolation coefficient for each of the grid directions that
+is uniform (`typeof(Δc) <: Number`) while it precomputes the ENO coefficients for reconstruction for all
+grid directions that are stretched. (After testing "on-the-fly" calculation of coefficients for stretched
+directions ended up being way too expensive and, therefore, is not supported.)
+
+```jldoctest
+julia> using Oceananigans
+
+julia> grid = grid = RectilinearGrid(size = (3, 4, 5), x = (0, 1), y = (0, 1), z = [-10, -9, -7, -4, -1.5, 0]);
+
+julia> WENO5(grid = grid)
+WENO5 advection scheme with:
+    ├── X regular
+    ├── Y regular
+    └── Z stretched
+```
+
+`WENO5(grid = grid, stretched_smoothness = true)` behaves similarly to `WENO5(grid = grid)` but, additionally,
+it also computes the smoothness indicators coefficients, ``β₀``, ``β₁``, and ``β₂``, taking into account
+the stretched dimensions.
+
+`WENO5(zweno = true)` implements a Z-WENO formulation for the WENO weights calculation
+
+Comments
+========
+
+All methods have the same execution speed. However, `stretched_smoothness = true` requires more memory (and
+slightly more computation time) and is not much impactful. As such, most of the times we urge users to use
+`WENO5(grid = grid)`, as this does not seem to decrease accuracy but does decreases memory utilization (and
+also results in a slight speed-up).
+
+(The above claims were made after some preliminary tests. Thus, we still users to perform some
+benchmarks/checks before performing, e.g., a large simulation on a "weirdly" stretched grid.)
+
+On the other hand, a Z-WENO formulation is *always* beneficial (also in case of a uniform mesh) with no major
+decrease in performance. The same can be said for the stretched `WENO5(grid = grid)` formulation in case of
+stretched grids.
+
+References
+==========
+
```

Comment:
@simone-silvestri, I didn't include all the info you provided in the PR comments in the docstring. I felt that what I included was helpful enough. But perhaps here we add 2-3 relevant references?

Reply:
@navidcy, thanks for the work. I'll add a couple of references